infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -641,12 +645,14 @@ protected void startMachines() throws InfrastructureException {
     // namespace.pods().watch(new AbnormalStopHandler());
     namespace.deployments().watchEvents(new MachineLogsPublisher());
     if (unrecoverableEventListenerFactory.isConfigured()) {
-      Map<String, Pod> pods = getContext().getEnvironment().getPods();
+      Set<String> toWatch = new HashSet<>();
+      KubernetesEnvironment environment = getContext().getEnvironment();
+      toWatch.addAll(environment.getPodsCopy().keySet());
+      toWatch.addAll(environment.getDeploymentsCopy().keySet());
```

Comment:
It's tricky, and a complication that this PR adds that I don't really like. As far as I can tell, the pod name specified in a deployment is ignored by OpenShift, and instead the pod just gets the deployment name + hashes. So this does work, it's just confusing. Since we're basically creating the deployment from the yaml directly, it means our recipes are beholden to how OpenShift/Kubernetes handle the objects themselves.

While double checking this, though, I found two issues that are now fixed: Pods weren't getting the deployment name label (accidentally deleted that line in refactoring) and Deployments weren't getting a unique name from the provisioner.